### PR TITLE
Support .gitignore exclude patterns i18

### PR DESCRIPTION
**Pull request by hadialqattan**

### CODE REVIEW

File: docs/CHANGELOG.md [hunk 1]
```diff
@@ -4,6 +4,7 @@ All notable changes to this project will be documented in this file.
 
 ## [Unreleased] - ././2020
 
+- [💪 Support gitignore exclude patterns by @hadialqattan](https://github.com/hakancelik96/unimport/pull/94)
```
Comment (hakancelik96) [inline docs/CHANGELOG.md:7]:
```suggestion
[💪 Support .gitignore exclude patterns by @hadialqattan](https://github.com/hakancelik96/unimport/pull/94)
```

File: docs/README.md [hunk 2]
```diff
@@ -295,6 +295,10 @@ import x
 
 ```
 
+## `.gitignore`
+
+To skip `.gitignore` glob patterns use `--gitignore` flag.
```
Comment (hakancelik96) [inline docs/README.md:300]: Can you specify that this feature is optional, if they want to use this feature, let's write that they need to download the **pathspec** library? As we did in **pyproject.toml** support.

File: setup.py [hunk 3]
```diff
@@ -30,7 +30,7 @@ def get_long_description():
     license_file="LICENSE",
     python_requires=">=3.6.0",
     packages=["unimport"],
-    install_requires=["libcst==0.3.10"],
+    install_requires=["libcst==0.3.10", "pathspec==0.8.0"],
```
Comment (hakancelik96) [inline setup.py:33]: Let's add this library to the extras_require section.
People who do not want to use this feature should not have to download this library.

File: docs/README.md (hunk 2 above)
Comment (hadialqattan) [reply docs/README.md:300]: I got your point!
I'm gonna work on that!

**Comment by hadialqattan** [comment]
If everything is fine, don't forget to link this PR with #18

**Comment by hakancelik96** [comment]
> If everything is fine, don't forget to link this PR with #18

It's not over yet, I will continue to review it when I have time.

File: unimport/__main__.py [hunk 4]
```diff
@@ -63,6 +70,19 @@ def show(
         )
 
 
+def get_gitignore(root: Union[Path, str]) -> PathSpec:
```
Comment (hakancelik96) [inline unimport/__main__.py:73]:
```suggestion
def get_gitignore() -> PathSpec:
```

File: unimport/__main__.py [hunk 5]
```diff
@@ -63,6 +70,19 @@ def show(
...
+    """Return a PathSpec matching gitignore content."""
+    if not PathSpec:
+        return None
+    lines: List[str] = []
+    path = os.path.join(root, ".gitignore")
```
Comment (hakancelik96) [inline unimport/__main__.py:78]:
```suggestion
    path: Path = Path(".gitignore")
```

File: unimport/__main__.py [hunk 6]
```diff
@@ -1,10 +1,17 @@
 import argparse
 import difflib
+import os
```
Comment (hakancelik96) [inline unimport/__main__.py:3]:
```suggestion
```

File: unimport/__main__.py [hunk 7]
```diff
@@ -63,6 +70,19 @@ def show(
         )
 
 
+def get_gitignore(root: Union[Path, str]) -> PathSpec:
+    """Return a PathSpec matching gitignore content."""
+    if not PathSpec:
+        return None
+    lines: List[str] = []
+    path = os.path.join(root, ".gitignore")
+    if os.path.isfile(path) and root:
```
Comment (hakancelik96) [inline unimport/__main__.py:79]:
```suggestion
    if path.is_file():
```

File: unimport/__main__.py [hunk 8]
```diff
@@ -63,6 +70,19 @@ def show(
...
+        if os.access(path, os.R_OK):
+            with tokenize.open(path) as ignore_file:
+                lines = ignore_file.readlines()
```
Comment (hakancelik96) [inline unimport/__main__.py:82]:
```suggestion
        lines: List[str] = tokenize.open(path).readlines()
```

File: unimport/__main__.py [hunk 9]
```diff
@@ -63,6 +70,19 @@ def show(
         )
 
 
+def get_gitignore(root: Union[Path, str]) -> PathSpec:
+    """Return a PathSpec matching gitignore content."""
+    if not PathSpec:
+        return None
+    lines: List[str] = []
```
Comment (hakancelik96) [inline unimport/__main__.py:77]:
```suggestion
```

File: unimport/__main__.py [hunk 10]
```diff
@@ -170,11 +195,17 @@ def main(argv: Optional[List[str]] = None) -> int:
         exclude_list.append(namespace.exclude)
     if hasattr(session.config, "exclude"):
         exclude_list.append(session.config.exclude)  # type: ignore
+    if namespace.gitignore:
+        gitignore = get_gitignore(Path("."))
+    else:
+        gitignore = get_gitignore("")
```
Comment (hakancelik96) [inline unimport/__main__.py:201]:
```suggestion
    if namespace.gitignore:
        gitignore = get_gitignore()
    else:
        gitignore = None
```

File: unimport/session.py [hunk 11]
```diff
@@ -58,11 +64,17 @@ def list_paths(
             file_names = start.glob(self.GLOB_PATTERN)
         else:
             file_names = [start]
-        yield from filter(
-            lambda filename: include_regex.search(str(filename))
-            and not exclude_regex.search(str(filename)),
-            file_names,
-        )
+
+        def matcher(filename: Path):
```
Comment (hakancelik96) [inline unimport/session.py:68]:
```suggestion
        def matcher(filename: Path) -> bool:
```

File: unimport/session.py [hunk 12]
```diff
@@ -4,6 +4,11 @@
 from pathlib import Path
 from typing import Iterable, Iterator, Optional, Tuple
 
+try:
+    from pathspec import PathSpec
+except ModuleNotFoundError:
+    PathSpec = None
```
Comment (hakancelik96) [inline unimport/session.py:10]:
```suggestion
    pass
```

**Review by hakancelik96 — CHANGES_REQUESTED** [review]
I wrote some suggestions, I hope I didn't make a mistake.

**Comment by hakancelik96** [comment]
It looks like you deleted the repo.

**Comment by hadialqattan** [comment]
> It looks like you deleted the repo.

Unfortunately!

**Comment by hadialqattan** [comment]
Anyway, I've moved these changes manually to another PR #102

File: unimport/__main__.py (hunk 9 above)
Comment (hadialqattan) [reply unimport/__main__.py:77]: This line shouldn't be removed, because `.gitignore` may not exist!

**Pull request closed without merging**